### PR TITLE
fix message length parsing on new hardware versions

### PR DESCRIPTION
**Pull request by Matthias-pixel**
I have two SEM6000 outlets with hardware versions 2 and 3. The library works fine with hardware version 2. With hardware version 3 I got the following stacktrace(dev in line 2 is my device):

```
  File "test.py", line 13, in connect
    dev.authorize(pin)
  File "/home/pi/git/fork-python3-voltcraft-sem6000/sem6000/sem6000.py", line 225, in authorize
    self._send_command(command)
  File "/home/pi/git/fork-python3-voltcraft-sem6000/sem6000/sem6000.py", line 155, in _send_command
    self._bluetooth_lowenergy_interface.write_to_characteristic(SEM6000.CHARACTERISTIC_UUID_CONTROL, encoded_command)
  File "/home/pi/git/fork-python3-voltcraft-sem6000/sem6000/bluetooth_lowenergy_interface/timeout_decorator.py", line 15, in decorated_function
    return_value = function(*s, **d)
  File "/home/pi/git/fork-python3-voltcraft-sem6000/sem6000/bluetooth_lowenergy_interface/bluepy_interface.py", line 141, in write_to_characteristic
    return characteristic.write(data, self._is_notifications_enabled)
  File "/home/pi/.local/lib/python3.7/site-packages/bluepy/btle.py", line 198, in write
    return self.peripheral.writeCharacteristic(self.valHandle, val, withResponse)
  File "/home/pi/.local/lib/python3.7/site-packages/bluepy/btle.py", line 541, in writeCharacteristic
    return self._getResp('wr')
  File "/home/pi/.local/lib/python3.7/site-packages/bluepy/btle.py", line 405, in _getResp
    resp = self._waitResp(wantType + ['ntfy', 'ind'], timeout)
  File "/home/pi/.local/lib/python3.7/site-packages/bluepy/btle.py", line 366, in _waitResp
    raise BTLEGattError("Bluetooth command failed", resp)
bluepy.btle.BTLEGattError: Bluetooth command failed (code: 3, error: Attribute can't be written)
```

This was fixed by only enabling notifications if the hardware version is below 3 and setting the MTU to 260 if not.
Since the SEM6000Delegate class needs to know the hardware version as well, it's only created after a connection is established.
As a last change, the length_of_payload is increased by 2 if the two fixed bytes at the end of a message aren't sent, as seen in API.md under Capture measurement.

### CODE REVIEW

File: sem6000/parser.py [hunk 1]
```diff
@@ -14,9 +14,11 @@ def __str__(self):
         return "message has invalid payload length for " + self.message_class.__name__ +  " (expected: " + str(self.expected_payload_length) + ", actual=" + str(self.actual_payload_length) + ")"
 
 class MessageParser:
-    def __init__(self, year_diff=None):
+    def __init__(self, hardware_version=None, year_diff=None):
         # the device only operates with two digit years
```
Comment (moormaster) [inline sem6000/parser.py]: These comments belong to the lines setting self.year_diff. Please place them directly before line 22.
```
# the device only operates with two digit years
# determine or set the difference to the current 4 digit year
```

File: sem6000/parser.py [hunk 2]
```diff
@@ -29,6 +31,9 @@ def _parse_payload(self, data):
         length_of_payload = data[1]
 
         payload = data[2:2+length_of_payload-1]
+        if (data[-2:len(data)] != bytes.fromhex("ffff")) and self.hardware_version == 3:
+            length_of_payload += 2
```
Comment (moormaster) [inline sem6000/parser.py:37]: The payload is increased after it was extracted from the notification. Isn't the last byte of the "total consumption" value missing in the return value?

Please also add a command about why the increase of the payload is necessary, i.e.
```
# In hardware version 3 the "capture measurement"-notification contains a
# payload length that is 2 bytes too short
if (data[-2:len(data)] != bytes.fromhex("ffff")) and self.hardware_version == 3:
    length_of_payload += 2

payload = data[2:2+length_of_payload-1]
```

File: sem6000/sem6000.py [hunk 3]
```diff
@@ -123,7 +124,13 @@ def _reconnect(self):
...
             raise e
 
-        self._bluetooth_lowenergy_interface.enable_notifications()
+        #get hardware version  
+        device_info = self._bluetooth_lowenergy_interface.read_from_characteristic(self.CHARACTERISTIC_UUID_VERSION)
+        self.hardware_version = int(device_info[13])
+        if self.hardware_version < 3:
+            self._bluetooth_lowenergy_interface.enable_notifications()
+        else:
+            self._bluetooth_lowenergy_interface._peripheral.setMTU(260)
```
Comment (moormaster) [inline sem6000/sem6000.py:133]: I think the MTU can be set to 260 also for the older devices. I have pushed the necessary changes in the upstream master branch and tested it with the hardware version 2 devices that I own. Please rebase your branch onto the most current commit of the upstream branch and test if that works you.

**Review by moormaster — COMMENTED** [review]
Hey,

thank you for your pull request. I have added some questions and comments. Please look at them and confirm the proposed changes.

File: sem6000/parser.py (hunk 2 above)
Comment (Matthias-pixel) [reply sem6000/parser.py:37]: Unfortunately I made this Pull request during an internship and no longer have the outlets. I wouldn't be able to test any changes to functionality.
All I can say is that it worked when I tested it.

File: sem6000/sem6000.py (hunk 3 above)
Comment (Matthias-pixel) [reply sem6000/sem6000.py:133]: I'm not that experienced with git and I think I messed it up (commits 503df8b and 0efd84a).
Should I delete the branch, redo the changes and create a new Pull request?

Comment (moormaster) [reply sem6000/sem6000.py:133]: Yeah if that is easier for you you should try it that way - Be sure to fetch the most recent changes into the master branch of your repository before creating the new branch. (see https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork )

**Comment by moormaster** [comment]
In the meantime I bought a new device which has hardware version 3. I have included your commits 

https://github.com/moormaster/python3-voltcraft-sem6000/commit/4c2f993a23836874a792ad3571f0e288849edfa5
https://github.com/moormaster/python3-voltcraft-sem6000/commit/4377088b29ef1bb75ad2d13786c898f9e5384a2d

together with a lot of other changes that were necessary to pass the integration test with a hw v3 device. I am therefore closing this issue. Thank you again for your contribution!

**Pull request closed without merging**